### ISSUE TITLE
Performance degrades as more files are open.

### ISSUE THREAD

**Issue by agorrod** [wiredtiger/wiredtiger#1865]
I made a change to the wtperf workload generator to have it open additional tables. The change causes handles to be created at the populate phase, and then a cursor is opened/closed when the workload begins.

I run the following configuration file:

```
conn_config="cache_size=1G,file_manager=(close_idle_time=1000)"
table_config="type=file"
icount=50000000
populate_threads=1
compact=false
threads=((count=8,read=1),(count=8,update=1))
run_time=120
report_interval=5
table_count=1
table_count_idle=10240
```

Switching between ```table_count_idle=0``` and ```table_count_idle=10240```. The performance results are:

| Phase | No Idle | 10k idle |
|---|---|---|--|--|
| Load (seconds) | 27s | 28s |
| Read rate (ops/sec) | 494324 | 126487 |
| Update rate (ops/sec) | 494324 | 126487 |

That means that having 10k tables open causes performance to drop by 4x - even when only a single table is being actively used.

This could become a bigger problem when we make changes to the sweep server so that it is much less aggressive at closing idle handles.

The changes to wtperf are in pull request #1864

**Comment by keithbostic** [comment]
I did a few runs on FreeBSD.

I started with the original CONFIG and then worked through a few different idle file counts:
```
        conn_config="cache_size=1G,file_manager=(close_idle_time=1000)"
        table_config="type=file"
        icount=50000000
        populate_threads=1
        compact=false
        threads=((count=8,read=1),(count=8,update=1))
        run_time=120
        report_interval=5
        table_count=1
        #table_count_idle=10240
 
        #table_count_idle=10240
        read: avg: 6193203
        insert: update: avg: 6191893
        
        table_count_idle=1024
        read: avg: 6035128
        insert: update: avg: 6035128

        table_count_idle=5120
        read: avg: 6326618
        insert: update: avg: 6326491
        
        table_count_idle=10240
        read: avg: 4742485
        insert: update: avg: 5034903
```
Seems like there's a change after 5K open handles.

**Comment by sueloverso** [comment]
A couple thoughts.  Once this is merged into develop we should commit that wtperf config and create a Jenkins job and plot this.  It would be good for the job to run multiple table_count_idle settings and plot them on the same chart - I'm happy to do that as I did something similar to in perf-log-prealloc.

Also this is probably related to #1439 and #1634 which also deal with long lists of open files.

**Comment by agorrod** [comment]
> commit that wtperf config and create a Jenkins job and plot this

That seems like a good idea. I've merged that pull request now.

> Also this is probably related to #1439 and #1634 which also deal with long lists of open files.

That's possible, though I'm surprised in this particular workload. What the wtperf code is doing is opening and closing cursors on a lot of tables, then opening a single cursor on an "active" table.

It proceeds to perform either reads or updates using that cursor. I've set the idle timeout long enough that sweep should not be running. I'm a bit stumped. I'm going to gather statistics from two runs to see if there is anything that could account for the difference.

I tried just opening file handles, rather that WiredTiger tables to see if it was something in the kernel, but doing that didn't trigger a performance difference up to 15k open files.

**Comment by agorrod** [comment]
My suspicion here is that the eviction server is spending a lot of time walking through the handles with no content. Comparing the workload phase between a run with no "idle handles", and one with 15k idle handles I'm seeing that the one with no idle handles has 5x as many "pages walked for eviction" as the one with many extra handles.

**Comment by agorrod** [comment]
If I increase the cache size so that this workload runs in memory (15G), then the performance difference between 0 and 15k idle handles is much less significant (within standard run-to-run deviation).

**Comment by sueloverso** [comment]
@agorrod I suspect that running a quick zoom profile might pinpoint the problem.  Send me the two commands or two configs that you're using.

**Comment by agorrod** [comment]
I tried Zoom - but I ran into AWS network access problems.

The config file I used was:

```
conn_config="cache_size=15G,file_manager=(close_idle_time=1000),statistics=[all],statistics_log=(wait=5)"
table_config="type=file"
icount=50000000
populate_threads=1
compact=false
threads=((count=8,read=1),(count=8,update=1))
run_time=240
report_interval=5
table_count=1
```

The command lines were:

```
rm -rf ./WT_TEST/* && ./bench/wtperf/wtperf -O ./multi.wtperf -o verbose=2 -o table_count_idle=0
```

and:

```
rm -rf ./WT_TEST/* && ./bench/wtperf/wtperf -O ./multi.wtperf -o verbose=2 -o table_count_idle=15000
```

**Comment by sueloverso** [comment]
Alex, what I do for AWS systems is run zoom on AWS on the command line and then `scp` the zoom file to my local machine and just view it there.  The command I use is:
`zoom run --thread_time_trigger 10 --process <pid> --time_limit 15`. That runs sampling every 10 msecs for 15 seconds on PID.

**Comment by sueloverso** [comment]
I should note that the zoom command is interactive, and you use keyboard controls to start or stop (and quit) the profiling at the time you think you want the data.  In interactive mode, you can start/stop multiple times during a single run and it'll generate a separate zoom file for each one.

**Comment by michaelcahill** [comment]
> I should note that the zoom command is interactive, and you use keyboard controls to start or stop (and quit) the profiling at the time you think you want the data. In interactive mode, you can start/stop multiple times during a single run and it'll generate a separate zoom file for each one.

FWIW, it’s also possible to tunnel in using “ssh -L 6666:localhost:6666 remote-host”, then connect the Zoom GUI to localhost:6666.

Michael.

**Comment by agorrod** [comment]
I generally setup an ssh tunnel - but it wasn't working the other day. I'm sure it was operator error. Running zoom locally from the command line is a nice alternative.
